go/marshal/decode.go

```diff
@@ -144,11 +155,29 @@ func uintDecoder(v types.Value, rv reflect.Value) {
 	}
 }
 
-var structDecoderCache struct {
+type decoderCacheT struct {
 	sync.RWMutex
 	m map[reflect.Type]decoderFunc
 }
 
+var decoderCache = &decoderCacheT{}
+
+func (c *decoderCacheT) get(t reflect.Type) decoderFunc {
+	c.RLock()
```

Comment:
You _could_ do this in 3 lines with
```
c.RLock()
defer c.RUnlock()
return c.m[t]
```
and using `defer` is probably a better habit anyway.